gnucash/import-export/csv-imp/gnc-csv-price-import-settings.cpp

```diff
@@ -0,0 +1,258 @@
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+\********************************************************************/
+/** @file gnc-csv-price-import-settings.cpp
+    @brief CSV Import Settings
+    @author Copyright (c) 2014 Robert Fewell
+    @author Copyright (c) 2016 Geert Janssens
+*/
+
+#include "gnc-csv-import-settings.hpp"
+#include "gnc-csv-price-import-settings.hpp"
+#include <sstream>
+
+extern "C"
+{
+#include <config.h>
+
+#include <gtk/gtk.h>
+#include <glib/gi18n.h>
+
+#include "Account.h"
+#include "gnc-state.h"
+#include "gnc-ui-util.h"
+}
+
+const std::string settings_type{"PRICE"};
+
+#define CSV_COL_TYPES    "ColumnTypes"
+
+#define CSV_TO_CURR      "PriceToCurrency"
+#define CSV_FROM_COMM    "PriceFromCommodity"
+
+G_GNUC_UNUSED static QofLogModule log_module = GNC_MOD_IMPORT;
+
+preset_vec_price presets_price;
+
+static std::shared_ptr<CsvPriceImpSettings> create_int_no_preset(void)
+{
+    auto preset = std::make_shared<CsvPriceImpSettings>();
+    preset->m_name = get_no_settings();
+    preset->m_settings_type = settings_type;
+
+    return preset;
+}
+
+static std::shared_ptr<CsvPriceImpSettings> create_int_gnc_exp_preset(void)
+{
+    auto preset = std::make_shared<CsvPriceImpSettings>();
+    preset->m_name = get_gnc_exp();
+    preset->m_skip_start_lines = 1;
+
+    /* FIXME date and currency format should still be aligned with export format!
+     * That's currently hard to do, because the export uses whatever the user
+     * had set as global preference.
+    preset->date_active = 0;
+    preset->currency_active = 0;
+    */
+    preset->m_column_types_price = {
+            GncPricePropType::DATE,
+            GncPricePropType::AMOUNT,
+            GncPricePropType::FROM_COMMODITY,
+            GncPricePropType::TO_CURRENCY
+    };
+    return preset;
+}
+
+/**************************************************
+ * find
+ *
+ * find all settings entries in the state key file
+ * based on settings type.
+ **************************************************/
+const preset_vec_price& get_import_presets_price (void)
+{
+    // Search all Groups in the state key file for ones starting with prefix
+    auto preset_names = std::vector<std::string>();
+    auto keyfile = gnc_state_get_current ();
+    gsize grouplength;
+    gchar **groups = g_key_file_get_groups (keyfile, &grouplength);
+
+    /* Start by building a sorted list of candidate presets as found in the state file */
+    for (gsize i=0; i < grouplength; i++)
+    {
+        auto group = std::string(groups[i]);
+        auto gp = get_prefix() + settings_type + " - ";
```

Comment:
In the template function and with a static get_group_prefix() function defined per derived class, this would become
```
auto gp = SType::get_group_prefix()
```

And move it outside of the for loop for efficiency.